src/Protocols/PacketStream/Base.hs

```diff
@@ -0,0 +1,213 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{- |
+Definitions and instances of the packet stream protocol
+-}
+module Protocols.PacketStream.Base (
+  PacketStreamM2S (..),
+  PacketStreamS2M (..),
+  PacketStream,
+  unsafeToPacketStream,
+  fromPacketStream,
+  forceResetSanity,
+  filterMetaS,
+  filterMeta,
+  mapMetaS,
+  mapMeta,
+) where
+
+import Clash.Prelude hiding (sample)
+import qualified Prelude as P
+
+import qualified Protocols.Df as Df
+import qualified Protocols.DfConv as DfConv
+import Protocols.Hedgehog.Internal
+import Protocols.Internal
+
+import Control.DeepSeq (NFData)
+import Data.Coerce (coerce)
+import qualified Data.Maybe as Maybe
+import Data.Proxy
+
+{- | Data sent from manager to subordinate, a simplified AXI4-Stream like interface
+with metadata that can only change on packet delineation.
+_tdest, _tuser and _tid are bundled into one big _meta field which holds metadata.
+There are no null or position bytes so _tstrb is replaced by a last indicator
+that indicates the index of the last valid byte in the _data vector.
+_tvalid is modeled via wrapping this in a `Maybe`.
+-}
+data PacketStreamM2S (dataWidth :: Nat) (metaType :: Type) = PacketStreamM2S
+  { _data :: Vec dataWidth (BitVector 8)
+  -- ^ The bytes to be transmitted
+  , _last :: Maybe (Index dataWidth)
+  -- ^ If Nothing, we are not yet at the last byte, otherwise index of last valid byte of _data
+  , _meta :: metaType
+  -- ^ the metadata of a packet. Must be constant during a packet.
+  , _abort :: Bool
+  -- ^ If True, the current transfer is aborted and the subordinate should ignore the current transfer
+  }
+  deriving (Eq, Generic, ShowX, Show, NFData, Bundle, Functor)
+
+{- | Data sent from the subordinate to the manager
+The only information transmitted is whether the subordinate is ready to receive data
+-}
+newtype PacketStreamS2M = PacketStreamS2M
+  { _ready :: Bool
+  -- ^ Iff True, the subordinate is ready to receive data
+  }
+  deriving (Eq, Generic, ShowX, Show, NFData, Bundle, NFDataX)
+
+-- | The packet stream protocol for communication between components
+data PacketStream (dom :: Domain) (dataWidth :: Nat) (metaType :: Type)
+
+deriving instance
+  (KnownNat dataWidth, NFDataX metaType) =>
+  NFDataX (PacketStreamM2S dataWidth metaType)
+
+instance Protocol (PacketStream dom dataWidth metaType) where
+  type
+    Fwd (PacketStream dom dataWidth metaType) =
+      Signal dom (Maybe (PacketStreamM2S dataWidth metaType))
+  type Bwd (PacketStream dom dataWidth metaType) = Signal dom PacketStreamS2M
```

Comment:
The flow control signals of `PacketStream` are isomorphic to those of `Df`.
And going on your explanation [here](https://github.com/clash-lang/clash-protocols/discussions/84#discussioncomment-9948153) I think the meaning of the flow control signals are basically the seems between the two protocols.
Would it make sense to reuse those `Df` data types in `PacketStream`?

```haskell
instance Protocol (PacketStream dom dataWidth metaType) where
  type Fwd (PacketStream dom dataWidth metaType) =
      Signal dom (Data (PacketStreamM2S dataWidth metaType))
  type Bwd (PacketStream dom dataWidth metaType) = Signal dom Ack
```

Reply:
Yes that's possible. We didn't do it because `AxiStream` protocol also didn't do it. We just followed that as our baseline.

Reply:
We will move over `Df` to use `Maybe`, see https://github.com/clash-lang/clash-protocols/issues/87